src/core/tui.rs

```diff
@@ -44,142 +38,72 @@ impl Tui {
-            ref current_view,
-            ..
-        } = *self;
-        info!("setting new terminal size");
         self.term_size = size;
-        if let Some(view) = views.get_mut(current_view) {
-            view.resize(size.1);
-        } else {
-            warn!("view {} not found", current_view);
-        }
-    }
-
-    pub fn open(&mut self, file_path: String) {
-        let client = self.client.clone();
-        let task = self.client
-            .new_view(Some(file_path.clone()))
-            .and_then(move |view_id| {
-                let view_client = ViewClient::new(client, view_id);
-                Ok((view_id, View::new(view_client, Some(file_path))))
-            });
-        self.pending_open_requests.push(Box::new(task));
+        self.editor.handle_resize(size);
     }
 
     fn exit(&mut self) {
         self.shutdown = true;
     }
 
-    fn handle_input(&mut self, event: Event) {
-        if Event::Key(Key::Ctrl('c')) == event {
-            self.exit()
-        } else if let Some(view) = self.views.get_mut(&self.current_view) {
-            view.handle_input(event)
+    pub fn handle_cmd(&mut self, cmd: Command) {
+        match cmd {
+            Command::Cancel => {
+                self.prompt = None;
+            },
+            Command::Quit => self.exit(),
+            Command::Save(view) => self.editor.save(view),
+            Command::Invalid(cmd) => {
+                error!("Received invalid editor command: {}", cmd);
+            }
         }
     }
 
-    pub fn set_theme(&mut self, theme: &str) {
-        let future = self.client.set_theme(theme).map_err(|_| ());
-        run(future);
-    }
-
-    fn process_open_requests(&mut self) {
-        if self.pending_open_requests.is_empty() {
-            return;
-        }
-
-        info!("process pending open requests");
-
-        let Tui {
-            ref mut pending_open_requests,
-            ref mut views,
-            ref mut current_view,
-            ref term_size,
-            ..
-        } = *self;
-
-        let mut done = vec![];
-        for (idx, task) in pending_open_requests.iter_mut().enumerate() {
-            match task.poll() {
-                Ok(Async::Ready((id, mut view))) => {
-                    info!("open request succeeded for {}", &id);
-                    done.push(idx);
-                    view.resize(term_size.1);
-                    views.insert(id, view);
-                    *current_view = id;
+    /// Global keybindings can be parsed here
+    fn handle_input(&mut self, event: Event) {
+        match event {
+            Event::Key(Key::Ctrl('c')) => self.exit(),
+            Event::Key(Key::Ctrl(':')) => {
+                if let Some(ref mut prompt) = self.prompt {
+                    prompt.handle_input(event);
+                } else {
+                    self.prompt = Some(CommandPrompt::default());
+                }
+            },
+            event => {
+                let mut out = None;
```

Comment:
Do you think this version would be more readable? I don't really like the gymnastic we have to do here to make the borrow checker happy:

```rust
            event => {
                // No command prompt is active, process the event normally.
                if self.prompt.is_none() {
                    self.editor.handle_input(event);
                    return;
                }

                // A command prompt is active.
                let mut prompt = self.prompt.take().unwrap();
                if let Some(cmd) = prompt.handle_input(event) {
                    // The event resulted in a command to process
                    self.handle_cmd(cmd);
                } else {
                    // Still not command, just update the prompt
                    self.prompt = Some(prompt);
                }
            }
```

It's a matter of taste though, (some people don't like unwraps for example) so I'll leave it to you :)

Reply:
I'm ok with unwrap in this case since it's guaranteed to be Ok. And yes I do like that better i wasn't very fond of splitting up the logic into two if let statements. i'll change it at once